libflux/src/semantic/tests.rs

```diff
@@ -0,0 +1,183 @@
+///
+/// A test case consists of:
+///
+/// 1. A name given as an identifier
+/// 2. An optional type environment (representing a prelude)
+/// 3. The flux program to be type checked
+/// 4. The expected types of the top-level identifiers of said program
+///
+/// # Example
+///
+/// ```
+/// test_infer! { instantiation,
+///     env: &[
+///         ("f", "forall [t0] where t0: Addable (a: t0, b: t0) -> t0"),
+///     ],
+///     src: "x = f",
+///     exp: &[
+///         ("x", "forall [t0] where t0: Addable (a: t0, b: t0) -> t0"),
+///     ],
+/// }
+/// ```
+///
+macro_rules! test_infer {
+    (
+        $name:ident,
+        $(env: $env:expr,)?
+        src: $src:expr,
+        exp: $exp:expr,
+    ) => {
+        #[test]
+        fn $name() {
+            let mut m = HashMap::new();
+            let mut f = Fresher::new();
+
+            $(
+                for (name, expr) in $env {
+                    let poly = parse(expr).unwrap().normalize(&mut f);
+                    m.insert(name.to_string(), poly);
+                }
+            )?
+
+            let pkg = parse_program($src);
+
+            let types = match nodes::infer_pkg_types(
+                &mut analyze(pkg, &mut f).unwrap(),
+                Environment::new(m.into()),
+                &mut f,
+            ) {
+                Err(err) => panic!("unexpected type error: {}", err.to_string()),
+                Ok((e, _)) => e.values,
+            };
+
+            let got: HashMap<String, PolyType> = types
+                .into_iter()
+                .map(|(name, poly)| (name, poly.normalize(&mut Fresher::new())))
+                .collect();
+
+            let want: HashMap<String, PolyType> = $exp
+                .iter()
+                .map(|(name, expr)| {
+                    (
+                        name.to_string(),
+                        parse(expr).unwrap().normalize(&mut Fresher::new()),
+                    )
+                })
+                .collect();
+
+            if want != got {
+                panic!("\n\n{}\n\n{}\n{}\n{}\n{}\n",
+                    "unexpected types:".red().bold(),
+                    "want:".green().bold(),
+                    want.iter().fold(String::new(), |acc, (name, poly)| acc
+                        + &format!("\t{}: {}\n", name, poly)),
+                    "got:".red().bold(),
+                    got.iter().fold(String::new(), |acc, (name, poly)| acc
+                        + &format!("\t{}: {}\n", name, poly)),)
+            }
+        }
+    };
+}
+
+/// The test_infer_err! macro generates test cases that don't type check.
+///
+/// These test cases consist of:
+///
+/// 1. A name given as an identifier
+/// 2. An optional type environment (representing a prelude)
+/// 3. A flux program that will not type check
+///
+/// # Example
+///
+/// ```
+/// test_infer_err! { undeclared variable,
+///     src: "x = f",
+/// }
+/// ```
+///
+macro_rules! test_infer_err {
```

Comment:
Do we want to include anything about the expected error?

Reply:
In the future yes. For right now I think it's a little premature until we formalize the structure of type errors. But we should (and we definitely will) have enough test coverage to ensure we don't have any false positives.

Reply:
Great, I am thinking something like this would be a good project soon https://github.com/elm/error-message-catalog/issues

Basically proof that error messages make sense in the context of their script.